### ISSUE TITLE
Default kind admin clusterrole is not enough to deploy a vcluster

### ISSUE THREAD

**Issue by rio** [loft-sh/vcluster#57]
I've already mentioned it in #42 but I thought I'd create a proper issue for it.

When trying to deploy a vcluster using a service account that has a namespaced rolebinding to the admin clusterrole the helm command fails because it tries to adds permissions that the admin clusterrole doesn't have. Below is the error output of `vcluster create`.

```
~ $ vcluster create vcluster-1
[info]   execute command: helm upgrade vcluster-1 vcluster --repo https://charts.loft.sh --version 0.3.0 --kubeconfig /tmp/247832214 --namespace vcluster --install --repository-config='' --values /tmp/285100285
[fatal]  error executing helm upgrade vcluster-1 vcluster --repo https://charts.loft.sh --version 0.3.0 --kubeconfig /tmp/247832214 --namespace vcluster --install --repository-config='' --values /tmp/285100285: Error: UPGRADE FAILED: failed to create resource: roles.rbac.authorization.k8s.io "vcluster-1" is forbidden: user "system:serviceaccount:vcluster:default" (groups=["system:serviceaccounts" "system:serviceaccounts:vcluster" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:[""], Resources:["configmaps"], Verbs:["*"]}
{APIGroups:[""], Resources:["endpoints"], Verbs:["*"]}
{APIGroups:[""], Resources:["events"], Verbs:["*"]}
{APIGroups:[""], Resources:["persistentvolumeclaims"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods/attach"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods/exec"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods/log"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods/portforward"], Verbs:["*"]}
{APIGroups:[""], Resources:["pods/proxy"], Verbs:["*"]}
{APIGroups:[""], Resources:["secrets"], Verbs:["*"]}
{APIGroups:[""], Resources:["services"], Verbs:["*"]}
{APIGroups:[""], Resources:["services/proxy"], Verbs:["*"]}
{APIGroups:["networking.k8s.io"], Resources:["ingresses"], Verbs:["*"]}
```

The admin clusterrole for kind v0.11.1 contains these permissions:

```
Name:         admin
Labels:       kubernetes.io/bootstrapping=rbac-defaults
Annotations:  rbac.authorization.kubernetes.io/autoupdate: true
PolicyRule:
  Resources                                       Non-Resource URLs  Resource Names  Verbs
  ---------                                       -----------------  --------------  -----
  rolebindings.rbac.authorization.k8s.io          []                 []              [create delete deletecollection get list patch update watch]
  roles.rbac.authorization.k8s.io                 []                 []              [create delete deletecollection get list patch update watch]
  configmaps                                      []                 []              [create delete deletecollection patch update get list watch]
  endpoints                                       []                 []              [create delete deletecollection patch update get list watch]
  persistentvolumeclaims                          []                 []              [create delete deletecollection patch update get list watch]
  pods                                            []                 []              [create delete deletecollection patch update get list watch]
  replicationcontrollers/scale                    []                 []              [create delete deletecollection patch update get list watch]
  replicationcontrollers                          []                 []              [create delete deletecollection patch update get list watch]
  services                                        []                 []              [create delete deletecollection patch update get list watch]
  daemonsets.apps                                 []                 []              [create delete deletecollection patch update get list watch]
  deployments.apps/scale                          []                 []              [create delete deletecollection patch update get list watch]
  deployments.apps                                []                 []              [create delete deletecollection patch update get list watch]
  replicasets.apps/scale                          []                 []              [create delete deletecollection patch update get list watch]
  replicasets.apps                                []                 []              [create delete deletecollection patch update get list watch]
  statefulsets.apps/scale                         []                 []              [create delete deletecollection patch update get list watch]
  statefulsets.apps                               []                 []              [create delete deletecollection patch update get list watch]
  horizontalpodautoscalers.autoscaling            []                 []              [create delete deletecollection patch update get list watch]
  cronjobs.batch                                  []                 []              [create delete deletecollection patch update get list watch]
  jobs.batch                                      []                 []              [create delete deletecollection patch update get list watch]
  daemonsets.extensions                           []                 []              [create delete deletecollection patch update get list watch]
  deployments.extensions/scale                    []                 []              [create delete deletecollection patch update get list watch]
  deployments.extensions                          []                 []              [create delete deletecollection patch update get list watch]
  ingresses.extensions                            []                 []              [create delete deletecollection patch update get list watch]
  networkpolicies.extensions                      []                 []              [create delete deletecollection patch update get list watch]
  replicasets.extensions/scale                    []                 []              [create delete deletecollection patch update get list watch]
  replicasets.extensions                          []                 []              [create delete deletecollection patch update get list watch]
  replicationcontrollers.extensions/scale         []                 []              [create delete deletecollection patch update get list watch]
  ingresses.networking.k8s.io                     []                 []              [create delete deletecollection patch update get list watch]
  networkpolicies.networking.k8s.io               []                 []              [create delete deletecollection patch update get list watch]
  poddisruptionbudgets.policy                     []                 []              [create delete deletecollection patch update get list watch]
  deployments.apps/rollback                       []                 []              [create delete deletecollection patch update]
  deployments.extensions/rollback                 []                 []              [create delete deletecollection patch update]
  localsubjectaccessreviews.authorization.k8s.io  []                 []              [create]
  pods/attach                                     []                 []              [get list watch create delete deletecollection patch update]
  pods/exec                                       []                 []              [get list watch create delete deletecollection patch update]
  pods/portforward                                []                 []              [get list watch create delete deletecollection patch update]
  pods/proxy                                      []                 []              [get list watch create delete deletecollection patch update]
  secrets                                         []                 []              [get list watch create delete deletecollection patch update]
  services/proxy                                  []                 []              [get list watch create delete deletecollection patch update]
  bindings                                        []                 []              [get list watch]
  events                                          []                 []              [get list watch]
  limitranges                                     []                 []              [get list watch]
  namespaces/status                               []                 []              [get list watch]
  namespaces                                      []                 []              [get list watch]
  persistentvolumeclaims/status                   []                 []              [get list watch]
  pods/log                                        []                 []              [get list watch]
  pods/status                                     []                 []              [get list watch]
  replicationcontrollers/status                   []                 []              [get list watch]
  resourcequotas/status                           []                 []              [get list watch]
  resourcequotas                                  []                 []              [get list watch]
  services/status                                 []                 []              [get list watch]
  controllerrevisions.apps                        []                 []              [get list watch]
  daemonsets.apps/status                          []                 []              [get list watch]
  deployments.apps/status                         []                 []              [get list watch]
  replicasets.apps/status                         []                 []              [get list watch]
  statefulsets.apps/status                        []                 []              [get list watch]
  horizontalpodautoscalers.autoscaling/status     []                 []              [get list watch]
  cronjobs.batch/status                           []                 []              [get list watch]
  jobs.batch/status                               []                 []              [get list watch]
  daemonsets.extensions/status                    []                 []              [get list watch]
  deployments.extensions/status                   []                 []              [get list watch]
  ingresses.extensions/status                     []                 []              [get list watch]
  replicasets.extensions/status                   []                 []              [get list watch]
  ingresses.networking.k8s.io/status              []                 []              [get list watch]
  poddisruptionbudgets.policy/status              []                 []              [get list watch]
  serviceaccounts                                 []                 []              [impersonate create delete deletecollection patch update get list watch]
```

You can reproduce this issue using the script below. You only require kind and kubectl in your path. Just write it to a file, give it the executable flag and execute it. It is idempotent so you can change any step and it wil just re run it if required.

```
#!/bin/bash

set -eux

if [ ! $(kind get clusters) ]; then
        kind create cluster --wait 1m
fi

# create a rolebinding to the default service account in the default namespace so our test pod
# has the admin role.
if [ ! $(kubectl get rolebinding namespace-admin) ]; then
        kubectl create rolebinding namespace-admin --serviceaccount default:default --clusterrole admin
fi

# start an ubuntu container that we can use as a base.
if [ ! $(kubectl get pod test-container) ]; then
        kubectl run test-container --image docker.io/library/ubuntu:20.04 -- sleep infinity
fi

kubectl wait --for condition=Ready pod test-container

# exec inside our test contaienr and install kubectl, helm and vcluster
kubectl exec -i test-container -- sh <<EOF
        set -eux

        apt-get update -qq
        apt-get install -qqy wget

        if [ ! -f /usr/local/bin/kubectl ]; then
                wget -qO /usr/local/bin/kubectl https://dl.k8s.io/release/v1.21.0/bin/linux/amd64/kubectl
                chmod +x /usr/local/bin/kubectl
        fi

        if [ ! -f /usr/local/bin/vcluster ]; then
                wget -qO- "https://github.com/loft-sh/vcluster/releases/latest" | sed -nE 's!.*"([^"]*vcluster-linux-amd64)".*!https://github.com\1!p' | xargs -n 1 wget -qO /usr/local/bin/vcluster
                chmod +x /usr/local/bin/vcluster
        fi

        if [ ! -f /usr/local/bin/helm ]; then
                wget -qO- https://get.helm.sh/helm-v3.6.1-linux-amd64.tar.gz | gzip -d | tar -O -x -f - linux-amd64/helm > /usr/local/bin/helm
                chmod +x /usr/local/bin/helm
        fi

        # test if the container can execute kubectl commands against the kind cluster
        kubectl get pod

        # print the privileges of the service account for debugging
        kubectl auth can-i --list

        # finally try to create a vcluster, this will fail!
        vcluster create my-failing-cluster
EOF
```

**Comment by rio** [comment]
Ok so got a chance to spend time to make it work with the test script I posted. Passing some extra helm values lets vcluster deploy.

```yaml
serviceAccount:
  create: false
  name: default

rbac:
  role:
    create: false
```

Now I'm not sure what the right course of action is here. I think that kind's admin clusterrole is the default one shipped using kubeadm. It being (I think) the maximum amount of privileges you can give a serviceaccount within a namespace, we should adjust the role that's created for vcluster accordingly.

**Comment by FabianKramm** [comment]
@rio thanks for creating this issue and detailed description. We just released a new version `v0.3.1-alpha.0` that uses specific verbs instead of `*` in the created rbac role. Could you check if that version fixes your problem?

**Comment by rio** [comment]
@FabianKramm, I can confirm it works now. Thanks for the fast fix! Loved the show on rawkode btw.
